pkg/apiserver/clusterinfo/host.go

```diff
@@ -0,0 +1,384 @@
+
+func loadHosts(db *gorm.DB) (HostMap, error) {
+	sql := "select TYPE, INSTANCE from INFORMATION_SCHEMA.CLUSTER_INFO;"
+	rows, err := querySQL(db, sql)
+	if err != nil {
+		return nil, err
+	}
+
+	hostMap := make(HostMap)
+	for _, row := range rows {
+		ip := parseIP(row[1])
+		var list []Instance
+		if instances, ok := hostMap[ip]; ok {
+			list = instances
+		}
+
+		list = append(list, Instance{
+			Address:    row[1],
+			ServerType: row[0],
+		})
+		hostMap[ip] = list
+	}
+
+	return hostMap, nil
+}
+
+type CPUCoreMap map[string]int
+
+func loadCPUCores(db *gorm.DB) (CPUCoreMap, error) {
+	sql := "select INSTANCE, VALUE from INFORMATION_SCHEMA.CLUSTER_HARDWARE where name = 'cpu-logical-cores';"
+	rows, err := querySQL(db, sql)
+	if err != nil {
+		return nil, err
+	}
+
+	var m = make(CPUCoreMap)
+	for _, row := range rows {
+		ip := parseIP(row[0])
+		cores, err := strconv.Atoi(row[1])
+		if err != nil {
+			continue
+		}
+		m[ip] = cores
+	}
+
+	return m, nil
+}
+
+func parseIP(addr string) string {
+	return strings.Split(addr, ":")[0]
+}
+
+type MemoryMap map[string]Memory
+
+func loadMemory(db *gorm.DB) (MemoryMap, error) {
+	sql := "select INSTANCE, NAME, VALUE from INFORMATION_SCHEMA.CLUSTER_LOAD where device_type = 'memory' and device_name = 'virtual';"
+	rows, err := querySQL(db, sql)
+	if err != nil {
+		return nil, err
+	}
+
+	var m = make(MemoryMap)
+	for _, row := range rows {
+		ip := parseIP(row[0])
+
+		var memory Memory
+		var ok bool
+		if memory, ok = m[ip]; !ok {
+			memory = Memory{}
+		}
+
+		switch row[1] {
+		case "total":
+			memory.Total, err = strconv.Atoi(row[2])
+			if err != nil {
+				continue
+			}
+		case "used":
+			memory.Used, err = strconv.Atoi(row[2])
+			if err != nil {
+				continue
+			}
+		default:
+			continue
+		}
+		m[ip] = memory
+	}
+
+	return m, nil
+}
+
+type CPUUsageMap map[string]CPUUsage
+
+func loadCPUUsage(db *gorm.DB) (CPUUsageMap, error) {
+	sql := "select INSTANCE, NAME, VALUE from INFORMATION_SCHEMA.CLUSTER_LOAD where device_type = 'cpu' and device_name = 'usage';"
+	rows, err := querySQL(db, sql)
+	if err != nil {
+		return nil, err
+	}
+
+	var m = make(CPUUsageMap)
+	for _, row := range rows {
+		ip := parseIP(row[0])
+
+		var cpu CPUUsage
+		var ok bool
+		if cpu, ok = m[ip]; !ok {
+			cpu = CPUUsage{}
+		}
+
+		switch row[1] {
+		case "system":
+			cpu.System, err = strconv.ParseFloat(row[2], 64)
+			if err != nil {
+				continue
+			}
+		case "idle":
+			cpu.Idle, err = strconv.ParseFloat(row[2], 64)
+			if err != nil {
+				continue
+			}
+		default:
+			continue
+		}
+		m[ip] = cpu
+	}
+
+	return m, nil
+}
+
+type PartitionMap map[string]Partition
+
+func queryPartition(db *gorm.DB, instance Instance) (PartitionMap, error) {
+	sql := fmt.Sprintf("select DEVICE_NAME, NAME, VALUE from INFORMATION_SCHEMA.CLUSTER_HARDWARE where type = '%s' and instance = '%s' and device_type = 'disk';",
+		instance.ServerType,
+		instance.Address,
+	)
+	rows, err := querySQL(db, sql)
```

Comment:
Your life will be easier if you use gorm as a ORM: https://gorm.io/docs/sql_builder.html#Run-Raw-SQL